Merge Marmalade V2

ggobugi27 opened the pull request:
Major Changes: 

- `poly-fungible-v3`: 
   - change `manifest:object{manifest}` to `uri:string`
   - Saves `policies[module{token-policy-v2}]` instead of `policy:moduel{token-policy-v2}`
   - Removes rotate
   - Enforces principal accounts 
   - Changes token protocol from `t:{manifest}` to `t:{(hash [token-details (at 'chain-id (chain-data))]))}`

- `token-policy-v2`: 
  - adds `enforce-withdraw`
  
- `concrete-policies`: 
   - Collection Policy: Emits collection events at init-collection, mint of collection tokens for tracking
   - Guard Policy: Adds guards per marmalade functions, optionally by providing guards at `read-msg`
   - Non-fungible Policy: Enforces the token to be non-fungible 
   - Royalty Policy: Saves creator's fungible account info and percentage, and pays percentage to the creator at every sale. 
  
 - `policy-manager`: 
    - Maps over token's policies and runs `enforce-**` functions 
    - Uses quote-manager and 
  
- `quote-manager` : 
  - saves quote information optionally by providing quote information with `read-msg`. 
  - Saves `seller-guard` which controls upgrading `quote-guards`
  - Saves `quote-guards`, which are list of guards which can upgrade quote information at `marmalade.buy`.
 
Pact Discussion Points :
- Circular Dependency:  
 In `quote-manager` and `policy-manager`, we create the tables `ledgers`, `policy-managers`, just to save its parent contract's module-capability-guards, so the functions can be protected. Other ideas were creating interfaces that include `enforce-ledger`. Please share thoughts. 
- `quote-manager.map-escrowed-buy` :    
 Policy manager calls `map-buy`, but to provide the escrow guard's capability, we call `map-escrowed-buy` in `quote-manager`. This means some `map-buy`'s are called withinin policy manager, but the `map-buy`'s that use escrowed accounts that pay fees to other policies, are called within quote managers

sirlensalot commented on pact/kip/poly-fungible-v3.pact:
```diff
@@ -211,35 +284,35 @@
   (defcap WITHDRAW:bool
-    (id:string seller:string amount:decimal timeout:integer sale-id:string)
+    (id:string seller:string amount:decimal timeout:time sale-id:string)
     @doc "Withdraws offer SALE from SELLER of AMOUNT of token ID after TIMEOUT."
```
Hmm why is this an event and not a managed cap?

sirlensalot commented on pact/ledger.pact:
```diff
@@ -141,38 +141,37 @@
 
-  (defun get-policy-info:object{policy-info} (id:string)
+  ;  Transform token-schema object to token-info object
+  (defun get-token-info:object{kip.token-policy-v2.token-info} (id:string)
```
Given that they're basically identical, do we need `policy-info` anymore?

sirlensalot commented on pact/ledger.pact:
```diff
@@ -190,42 +189,62 @@
-    (enforce-verify-manifest manifest)
-    (format "t:{}" [(at 'hash manifest)])
+  (defun create-token-id:string (token-details:object{token-details})
+    (format "t:{}"
```
Document format

sirlensalot commented on pact/root/coin.pact:
```diff
@@ -464,23 +508,24 @@
-            (format "Unrecognized reserved protocol: {}" [r]))))))
+    (if (validate-principal guard account)
+      true
+      (let ((r (check-reserved account)))
```
Unless there is a compelling reason we need to remove `check-reserved` and this code and only enforce `validate-principal`

sirlensalot commented on pact/ledger.pact:
```diff
@@ -293,19 +300,19 @@
+        "sender cannot be the receiver of a transfer")
+      (enforce-valid-transfer sender receiver (precision id) amount)
+      (with-capability (TRANSFER id sender receiver amount)
+        (enforce-transfer-policy id sender receiver amount)
```
Move out of cap acquire

sirlensalot commented on pact/ledger.pact:
```diff
@@ -314,41 +321,39 @@
+              {'account: "", 'previous: 0.0, 'current: 0.0})
+          )
+          (emit-event (RECONCILE id amount sender receiver))
+          (update-supply id amount)
```
This btw is the sensitive call requiring policies to NOT be in operation cap

sirlensalot commented on pact/ledger.pact:
```diff
@@ -314,41 +321,39 @@
+    (with-capability (LEDGER)
+      (with-capability (BURN id account amount)
+        (let ((token (get-token-info id)))
+          (marmalade.policy-manager.enforce-burn token account amount))
```
move out of BURN cap

sirlensalot reviewed (commented):
Partial review, still need to do `policies` `policy-manager` `quote-manager` folders

sirlensalot commented on pact/concrete-policies/collection-policy/collection-policy-v1.pact:
```diff
@@ -0,0 +1,172 @@
+    true)
+
+  (defun enforce-ledger:bool ()
+    (enforce-guard (marmalade.ledger.ledger-guard))
```
Looks like its a cap guard, all good

sirlensalot commented on pact/ledger.pact:
```diff
@@ -514,20 +523,19 @@
+      timeout:time
     )
     (step-with-rollback
       (with-capability (SALE id seller amount timeout (pact-id))
```
ie

```pact
(defpact sale ...
  (step-with-rollback 
    (with-capability (LEDGER)
      (marmalade.policy-manager.enforce-offer (get-token-info id) seller amount (pact-id)))
      (with-capability (SALE id seller amount timeout (pact-id)) ...))
    (with-capability (LEDGER)
      ...))
   (step
      (with-capability (LEDGER)
     ...)))
```

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+  )
+
+  (deftable ledgers:{ledger}
+    @doc "Designed to save one ledger in a single row")
```
```suggestion
    @doc "Singleton table for ledger guard storage")
```

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+  )
+
+  (defcap CONCRETE_POLICY:bool (policy-field:string policy:module{kip.token-policy-v2})
+    @doc "Event emission purpose"
```
fix docs or remove

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+  (defconst CONCRETE_POLICY_LIST:[string]
+    [NON_FUNGIBLE_POLICY ROYALTY_POLICY COLLECTION_POLICY GUARD_POLICY] )
+
+  (defun write-concrete-policy:bool (policy-field:string policy:module{kip.token-policy-v2})
```
I only see this in repl files. How is this installed? I don't see it in yaml files.

ggobugi27 replied:
I've added [pact/policy-manager/manager-init.pact](https://github.com/kadena-io/marmalade/blob/4bf65aa1646a6038ba634f814a789e868efc5ae4/pact/policy-manager/manager-init.pact)

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+    )
+    (enforce-ledger)
+    (with-capability (POLICY_MANAGER)
+      (let ((policies:[module{kip.token-policy-v2}]  (at 'policies token)))
```
`enforce-init` doesn't need a let, can these do the same? (there is a small cost for `let`)

ggobugi27 replied:
yes

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+    )
+    (enforce-ledger)
+    (with-capability (POLICY_MANAGER)
+      (let ((policies:[module{kip.token-policy-v2}]  (at 'policies token)))
```
same question as `enforce-mint`

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+
+       (with-capability (ESCROW sale-id)
+         ;; Run policies::enforce-buy
+         (map-buy token seller buyer buyer-guard amount sale-id policies)
```
move out of ESCROW cap

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+
+  (defun exists-quote:bool (sale-id:string)
+    @doc "Looks up quote table for quote"
+    (= (take 6 (typeof (try false (get-quote-info sale-id)))) "object")
```
For these kinds of `try`, you can simply use a let block with the potentially-failing code and return a boolean directly.

```pact
(try false (let ((q (get-quote-info sale-id))) true))
```

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+
+  (defun exists-msg-decimal:bool (msg:string)
+    @doc "Checks env-data field and see if the msg is a decimal"
+    (= (typeof  (try false (read-decimal msg))) "decimal")
```
```suggestion
    (try false (let ((d (read-decimal msg))) true))
```

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+  )
+
+  (defun exists-msg-object:bool (msg:string)
+    @doc "Checks env-data field and see if the msg is a object"
```
We're moving away from untyped objects, and this function is only used once, so suggest rename to `exists-msg-quote` and use suggestion below.

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+
+  (defun exists-msg-object:bool (msg:string)
+    @doc "Checks env-data field and see if the msg is a object"
+    (= (take 6 (typeof  (try false  (read-msg msg)))) "object")
```
We can leverage Pact's runtime typechecker to fail if it's not the right object type, and use the let block technique with that.
```suggestion
    (try false (let ((o:object{quote-spec} (read-msg msg))) true)
```

sirlensalot commented on pact/policy-manager/policy-manager.pact:
```diff
@@ -0,0 +1,339 @@
+  (policy::enforce-init token))
+
+ (defun map-init (token:object{token-info} policy-list:[module{kip.token-policy-v2}])
+  (map (token-init token) policy-list))
```
Not necessary but consider using a lambda instead of having to have `token-init` toplevel function:
```suggestion
  (map (lambda (p:module{kip.token-policy-v2}) (p::enforce-init token)) policy-list)
```

sirlensalot replied:
If you like this apply to other `map-XXX` functions. You can also let bind the lambda if you like that style better:


```pact
(let ((lam (lambda (p:module{kip.token-policy-v2}) 
             (p::enforce-XXX foo bar)))
  (map lam policy-list))
```

sirlensalot commented on pact/policy-manager/policy-manager.repl:
```diff
@@ -0,0 +1,330 @@
+   , 'upgrade: false })
+  (load "../quote-manager/quote-manager.pact")
+  (load "../policy-manager/policy-manager.pact")
+  (load "../ledger.pact")
```
typecheck policy manager

sirlensalot commented on pact/quote-manager/quote-manager.pact:
```diff
@@ -0,0 +1,211 @@
+      token-id:string
+      spec:object{quote-spec}
+    )
+    @doc "For event emission purposes"
```
remove docs like this

sirlensalot commented on pact/quote-manager/quote-manager.pact:
```diff
@@ -0,0 +1,211 @@
+  (deftable quotes:{quote-schema})
+
+  (defcap UPDATE_QUOTE:bool (sale-id:string)
+    @doc "Enforces quote-guad on update-quote"
```
```suggestion
    @doc "Enforces quote-guards on update-quote-price"
```

sirlensalot commented on pact/quote-manager/quote-manager.pact:
```diff
@@ -0,0 +1,211 @@
+
+  (deftable quotes:{quote-schema})
+
+  (defcap UPDATE_QUOTE:bool (sale-id:string)
```
```suggestion
  (defcap UPDATE_QUOTE_PRICE:bool (sale-id:string)
```
to distinguish from UPDATE_QUOTE_GUARD below

sirlensalot commented on pact/quote-manager/quote-manager.repl:
```diff
@@ -0,0 +1,144 @@
+;;load policy manager, ledger
+(load "../policy-manager/policy-manager.repl")
+
```
typecheck quote manager

sirlensalot reviewed (commented):
done! :)

ggobugi27 commented on pact/policies/fixed-issuance-policy/fixed-issuance-policy-v1.pact:
```diff
@@ -0,0 +1,125 @@
+  )
+
+  (defun enforce-ledger:bool ()
+     (enforce-guard (marmalade.ledger.ledger-guard))
```
yes, `marmalade-v2`. Changed it.

ggobugi27 commented on pact/policies/migration-policy/migration-policy.md:
```diff
@@ -0,0 +1,84 @@
+
+Note that `burn` need to be enabled in marmalade v1 policy for the desired token.
+
+#### Remove `migration-policy-v1` from the `adjustable-policies`
```
Moved migration policy out for this release, will add later

ggobugi27 commented:
Additional Changes : 
`poly-fungible-v3`
  - Remove ties to `policies` (`create-token`, `TOKEN`)  #142 

`token-policy-v2`: 
  - Add `timeout` argument to enforce-offer/buy/withdraw 

`ledger`
 - Add `creation-guard` in `create-token-id`  enforce `creation-guard` at `create-token` to prevent frontrun of 
token-id’s #142 
 - Added option for not using timeout, and signing to withdraw #119 
 - Added `ledger.interface.pact` to provide capabilities to grant policy-manager #132 
 - Changed `timeout` to integer, representing timestamps 
`policy-manager`
- Added `policy-manager.interface.pact` to provide capabilities to grant quote-manager #132 

 `royalty-policy-v1`
 - Restricts to only support `coin`

`collection-policy-v1`
 - Added enforcement of operator guard at `create-collection` to prevent frontrun #133 

Others: 
- removed example contracts to `v2-examples` (`basic-bidding-contract`, `fixed-issuance-policy`, `onchain-manifest-policy`)
- Refined caps to stream data more efficiently, renamed to use `-` instead of `_` #146 . 
Discussion point: left the caps name with underscores in `ledger`  to keep consistent from maramalade-v1
